README

========
paramiko
========

:Paramiko: Python SSH module
:Copyright: Copyright (c) 2003-2009  Robey Pointer <robeypointer@gmail.com>
:Copyright: Copyright (c) 2014  Jeff Forcier <jeff@bitprophet.org>
:License: LGPL
:Homepage: https://github.com/paramiko/paramiko/
:API docs: http://docs.paramiko.org


What
----

"paramiko" is a combination of the esperanto words for "paranoid" and
"friend".  it's a module for python 2.5+ that implements the SSH2 protocol
for secure (encrypted and authenticated) connections to remote machines.
unlike SSL (aka TLS), SSH2 protocol does not require hierarchical
certificates signed by a powerful central authority. you may know SSH2 as
the protocol that replaced telnet and rsh for secure access to remote
shells, but the protocol also includes the ability to open arbitrary
channels to remote services across the encrypted tunnel (this is how sftp
works, for example).

it is written entirely in python (no C or platform-dependent code) and is
released under the GNU LGPL (lesser GPL). 

the package and its API is fairly well documented in the "doc/" folder
that should have come with this archive.


Requirements
------------

  - python 2.5 or better <http://www.python.org/>
  - pycrypto 2.1 or better <https://www.dlitz.net/software/pycrypto/>
  - ecdsa 0.9 or better <https://pypi.python.org/pypi/ecdsa>

If you have setuptools, you can build and install paramiko and all its
dependencies with this command (as root)::

    easy_install ./


Portability
-----------

i code and test this library on Linux and MacOS X. for that reason, i'm
pretty sure that it works for all posix platforms, including MacOS. it
should also work on Windows, though i don't test it as frequently there.
if you run into Windows problems, send me a patch: portability is important
to me.

some python distributions don't include the utf-8 string encodings, for
reasons of space (misdirected as that is). if your distribution is
missing encodings, you'll see an error like this::

    LookupError: no codec search functions registered: can't find encoding

this means you need to copy string encodings over from a working system.
(it probably only happens on embedded systems, not normal python
installs.) Valeriy Pogrebitskiy says the best place to look is
``.../lib/python*/encodings/__init__.py``.


Bugs & Support
--------------

Please file bug reports at https://github.com/paramiko/paramiko/. There is currently no mailing list but we plan to create a new one ASAP.


Kerberos Support
----------------

If you want paramiko to do kerberos authentication using GSS-API or SSPI, you 
need the following python packages:

- pyasn1 0.1.7 or better
- python-gssapi 0.4.0 or better (Unix)
- pywin32 2.1.8 or better (Windows)

So you have to install pyasn1 and python-gssapi on Unix or pywin32 on Windows.
To enable GSS-API / SSPI authentication see the demos or paramiko docs.
Note: If you use Microsoft SSPI for kerberos authentication and credential 
delegation in paramiko, make sure that the target host is trusted for 
delegation in the active directory configuration. For details see:
http://technet.microsoft.com/en-us/library/cc738491%28v=ws.10%29.aspx


Demo
----

several demo scripts come with paramiko to demonstrate how to use it.
probably the simplest demo of all is this::

    import paramiko, base64
    key = paramiko.RSAKey(data=base64.decodestring('AAA...'))
    client = paramiko.SSHClient()
    client.get_host_keys().add('ssh.example.com', 'ssh-rsa', key)
    client.connect('ssh.example.com', username='strongbad', password='thecheat')
    stdin, stdout, stderr = client.exec_command('ls')
    for line in stdout:
        print '... ' + line.strip('\n')
    client.close()

...which prints out the results of executing ``ls`` on a remote server.
(the host key 'AAA...' should of course be replaced by the actual base64
encoding of the host key.  if you skip host key verification, the
connection is not secure!)

the following example scripts (in demos/) get progressively more detailed:

:demo_simple.py:
    calls invoke_shell() and emulates a terminal/tty through which you can
    execute commands interactively on a remote server.  think of it as a
    poor man's ssh command-line client.

:demo.py:
    same as demo_simple.py, but allows you to authenticiate using a
    private key, attempts to use an SSH-agent if present, and uses the long
    form of some of the API calls.

:forward.py:
    command-line script to set up port-forwarding across an ssh transport.
    (requires python 2.3.)

:demo_sftp.py:
    opens an sftp session and does a few simple file operations.

:demo_server.py:
    an ssh server that listens on port 2200 and accepts a login for
    'robey' (password 'foo'), and pretends to be a BBS.  meant to be a
    very simple demo of writing an ssh server.

:demo_keygen.py:
    an key generator similar to openssh ssh-keygen(1) program with
    paramiko keys generation and progress functions.

Use
---

the demo scripts are probably the best example of how to use this package.
there is also a lot of documentation, generated with epydoc, in the doc/
folder.  point your browser there.  seriously, do it.  mad props to
epydoc, which actually motivated me to write more documentation than i
ever would have before.

there are also unit tests here::

    $ python ./test.py

which will verify that most of the core components are working correctly.


NEWS

====
NEWS
====

Highlights of what's new in each release.

Issues noted as "'ssh' #NN" can be found at https://github.com/bitprophet/ssh/.

Issues noted as "Fabric #NN" can be found at https://github.com/fabric/fabric/.


Kerberos Support
================

Paramiko supports now GSS-API / SSPI authentication and key exchange.
This feature requires the following Python packages:
- pyasn1 0.1.7 or better
- python-gssapi 0.4.0 or better (Unix)
- pywin32 2.1.8 or better (Windows)


Releases
========

v1.12.1 (8th Jan 2014)
----------------------

* #176: Fix AttributeError bugs in known_hosts file (re)loading. Thanks to
  Nathan Scowcroft for the patch & Martin Blumenstingl for the initial test
  case.
* #225: Note ecdsa requirement in README. Thanks to Amaury Rodriguez for the
  catch.

v1.11.3 (8th Jan 2014)
----------------------

* #176: Fix AttributeError bugs in known_hosts file (re)loading. Thanks to
  Nathan Scowcroft for the patch & Martin Blumenstingl for the initial test
  case.

v1.10.5 (8th Jan 2014)
----------------------

* #176: Fix AttributeError bugs in known_hosts file (re)loading. Thanks to
  Nathan Scowcroft for the patch & Martin Blumenstingl for the initial test
  case.

v1.12.0 (27th Sep 2013)
-----------------------

* #152: Add tentative support for ECDSA keys. *This adds the ecdsa
  module as a new dependency of Paramiko.* The module is available at
  [warner/python-ecdsa on Github](https://github.com/warner/python-ecdsa) and
  [ecdsa on PyPI](https://pypi.python.org/pypi/ecdsa).
    * Note that you might still run into problems with key negotiation --
      Paramiko picks the first key that the server offers, which might not be
      what you have in your known_hosts file.
    * Mega thanks to Ethan Glasser-Camp for the patch.
* #136: Add server-side support for the SSH protocol's 'env' command. Thanks to
  Benjamin Pollack for the patch.

v1.11.2 (27th Sep 2013)
-----------------------

* #156: Fix potential deadlock condition when using Channel objects as sockets
  (e.g. when using SSH gatewaying). Thanks to Steven Noonan and Frank Arnold
  for catch & patch.
* #179: Fix a missing variable causing errors when an ssh_config file has a
  non-default AddressFamily set. Thanks to Ed Marshall & Tomaz Muraus for catch
  & patch.
* #200: Fix an exception-causing typo in `demo_simple.py`. Thanks to Alex
  Buchanan for catch & Dave Foster for patch.
* #199: Typo fix in the license header cross-project. Thanks to Armin Ronacher
  for catch & patch.

v1.10.4 (27th Sep 2013)
-----------------------

* #179: Fix a missing variable causing errors when an ssh_config file has a
  non-default AddressFamily set. Thanks to Ed Marshall & Tomaz Muraus for catch
  & patch.
* #200: Fix an exception-causing typo in `demo_simple.py`. Thanks to Alex
  Buchanan for catch & Dave Foster for patch.
* #199: Typo fix in the license header cross-project. Thanks to Armin Ronacher
  for catch & patch.

v1.11.1 (20th Sep 2013)
-----------------------

* #162: Clean up HMAC module import to avoid deadlocks in certain uses of
  SSHClient. Thanks to Gernot Hillier for the catch & suggested
  fix.
* #36: Fix the port-forwarding demo to avoid file descriptor errors. Thanks to
  Jonathan Halcrow for catch & patch.
* #168: Update config handling to properly handle multiple 'localforward' and
  'remoteforward' keys. Thanks to Emre Yılmaz for the patch.

v1.10.3 (20th Sep 2013)
-----------------------

* #162: Clean up HMAC module import to avoid deadlocks in certain uses of
  SSHClient. Thanks to Gernot Hillier for the catch & suggested
  fix.
* #36: Fix the port-forwarding demo to avoid file descriptor errors. Thanks to
  Jonathan Halcrow for catch & patch.
* #168: Update config handling to properly handle multiple 'localforward' and
  'remoteforward' keys. Thanks to Emre Yılmaz for the patch.

v1.11.0 (26th Jul 2013)
-----------------------

* #98: On Windows, when interacting with the PuTTY PAgeant, Paramiko now
  creates the shared memory map with explicit Security Attributes of the user,
  which is the same technique employed by the canonical PuTTY library to avoid
  permissions issues when Paramiko is running under a different UAC context
  than the PuTTY Ageant process. Thanks to Jason R. Coombs for the patch.
* #100: Remove use of PyWin32 in `win_pageant` module. Module was already
  dependent on ctypes for constructing appropriate structures and had ctypes
  implementations of all functionality. Thanks to Jason R. Coombs for the
  patch.
* #87: Ensure updates to `known_hosts` files account for any updates to said
  files after Paramiko initially read them. (Includes related fix to guard
  against duplicate entries during subsequent `known_hosts` loads.) Thanks to
  `@sunweaver` for the contribution.

v1.10.2 (26th Jul 2013)
-----------------------

* #153, #67: Warn on parse failure when reading known_hosts file. Thanks to
  `@glasserc` for patch.
* #146: Indentation fixes for readability. Thanks to Abhinav Upadhyay for catch
  & patch.

v1.10.1 (5th Apr 2013)
----------------------

* #142: (Fabric #811) SFTP put of empty file will still return the attributes
  of the put file. Thanks to Jason R. Coombs for the patch.
* #154: (Fabric #876) Forwarded SSH agent connections left stale local pipes
  lying around, which could cause local (and sometimes remote or network)
  resource starvation when running many agent-using remote commands. Thanks to
  Kevin Tegtmeier for catch & patch.

v1.10.0 (1st Mar 2013)
--------------------

* #66: Batch SFTP writes to help speed up file transfers. Thanks to Olle
  Lundberg for the patch.
* #133: Fix handling of window-change events to be on-spec and not
  attempt to wait for a response from the remote sshd; this fixes problems with
  less common targets such as some Cisco devices. Thanks to Phillip Heller for
  catch & patch.
* #93: Overhaul SSH config parsing to be in line with `man ssh_config` (& the

paramiko/__init__.py
# Copyright (C) 2003-2011  Robey Pointer <robeypointer@gmail.com>
#
# This file is part of paramiko.
#
# Paramiko is free software; you can redistribute it and/or modify it under the
# terms of the GNU Lesser General Public License as published by the Free
# Software Foundation; either version 2.1 of the License, or (at your option)
# any later version.
#
# Paramiko is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with Paramiko; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.

"""
I{Paramiko} (a combination of the esperanto words for "paranoid" and "friend")
is a module for python 2.5 or greater that implements the SSH2 protocol for
secure (encrypted and authenticated) connections to remote machines.  Unlike
SSL (aka TLS), the SSH2 protocol does not require hierarchical certificates
signed by a powerful central authority.  You may know SSH2 as the protocol that
replaced C{telnet} and C{rsh} for secure access to remote shells, but the
protocol also includes the ability to open arbitrary channels to remote
services across an encrypted tunnel.  (This is how C{sftp} works, for example.)

The high-level client API starts with creation of an L{SSHClient} object.
For more direct control, pass a socket (or socket-like object) to a
L{Transport}, and use L{start_server <Transport.start_server>} or
L{start_client <Transport.start_client>} to negoatite
with the remote host as either a server or client.  As a client, you are
responsible for authenticating using a password or private key, and checking
the server's host key.  I{(Key signature and verification is done by paramiko,
but you will need to provide private keys and check that the content of a
public key matches what you expected to see.)}  As a server, you are
responsible for deciding which users, passwords, and keys to allow, and what
kind of channels to allow.

Once you have finished, either side may request flow-controlled L{Channel}s to
the other side, which are python objects that act like sockets, but send and
receive data over the encrypted session.

Paramiko is written entirely in python (no C or platform-dependent code) and is
released under the GNU Lesser General Public License (LGPL).

Website: U{https://github.com/paramiko/paramiko/}

Mailing list: U{paramiko@librelist.com<mailto:paramiko@librelist.com>}
"""

import sys

if sys.version_info < (2, 5):
    raise RuntimeError('You need python 2.5+ for this module.')


__author__ = "Jeff Forcier <jeff@bitprophet.org>"
__version__ = "1.12.0"
__license__ = "GNU Lesser General Public License (LGPL)"


from transport import SecurityOptions, Transport
from client import SSHClient, MissingHostKeyPolicy, AutoAddPolicy, RejectPolicy, WarningPolicy
from auth_handler import AuthHandler
from ssh_gss import SSH_GSSAuth
from channel import Channel, ChannelFile
from ssh_exception import SSHException, PasswordRequiredException, \
    BadAuthenticationType, ChannelException, BadHostKeyException, \
    AuthenticationException, ProxyCommandFailure
from server import ServerInterface, SubsystemHandler, InteractiveQuery
from rsakey import RSAKey
from dsskey import DSSKey
from ecdsakey import ECDSAKey
from sftp import SFTPError, BaseSFTP
from sftp_client import SFTP, SFTPClient
from sftp_server import SFTPServer
from sftp_attr import SFTPAttributes
from sftp_handle import SFTPHandle
from sftp_si import SFTPServerInterface
from sftp_file import SFTPFile
from message import Message
from packet import Packetizer
from file import BufferedFile
from agent import Agent, AgentKey
from pkey import PKey
from hostkeys import HostKeys
from config import SSHConfig
from proxy import ProxyCommand

# fix module names for epydoc
for c in locals().values():
    if issubclass(type(c), type) or type(c).__name__ == 'classobj':
        # classobj for exceptions :/
        c.__module__ = __name__
del c

from common import AUTH_SUCCESSFUL, AUTH_PARTIALLY_SUCCESSFUL, AUTH_FAILED, \
     OPEN_SUCCEEDED, OPEN_FAILED_ADMINISTRATIVELY_PROHIBITED,  OPEN_FAILED_CONNECT_FAILED, \
     OPEN_FAILED_UNKNOWN_CHANNEL_TYPE, OPEN_FAILED_RESOURCE_SHORTAGE

from sftp import SFTP_OK, SFTP_EOF, SFTP_NO_SUCH_FILE, SFTP_PERMISSION_DENIED, SFTP_FAILURE, \
     SFTP_BAD_MESSAGE, SFTP_NO_CONNECTION, SFTP_CONNECTION_LOST, SFTP_OP_UNSUPPORTED

from common import io_sleep

__all__ = [ 'Transport',
            'SSHClient',
            'MissingHostKeyPolicy',
            'AutoAddPolicy',
            'RejectPolicy',
            'WarningPolicy',
            'SecurityOptions',
            'SubsystemHandler',
            'Channel',
            'PKey',
            'RSAKey',
            'DSSKey',
            'Message',
            'SSHException',
            'AuthenticationException',
            'PasswordRequiredException',
            'BadAuthenticationType',
            'ChannelException',
            'BadHostKeyException',
            'ProxyCommand',
            'ProxyCommandFailure',
            'SFTP',
            'SFTPFile',
            'SFTPHandle',
            'SFTPClient',
            'SFTPServer',
            'SFTPError',
            'SFTPAttributes',
            'SFTPServerInterface',
            'ServerInterface',
            'BufferedFile',
            'Agent',
            'AgentKey',
            'HostKeys',
            'SSHConfig',
            'util',
            'io_sleep' ]
